mipengine/controller/celery_app.py

```diff
@@ -60,145 +56,75 @@ def __init__(
 class CeleryWrapper:
     def __init__(self, socket_addr: str):
         self._socket_addr = socket_addr
-        self._check_broker_connection_and_reset_celery_lock = Lock()
+        self._change_app_lock = Lock()
         self._celery_app = self._instantiate_celery_object()
 
-    def _close(self):
-        self._celery_app.close()
+    def _set_new_cel_app(self, cel_app: Celery):
+        with self._change_app_lock:
+            old_app = self._celery_app
+            self._celery_app = cel_app
+            old_app.close()
```

Comment:
Passing the celery object is a remnant from a previous implementation, will fix it.

Closing though the celery app before changing it, could lead in some very edge case scenario for a "queue task" to use the old celery app that is closed. Closing the app doesn't mean very much though but it could actually have an impact with a future celery version.

Reply:
it should be impossible for a queue task to happen on the old celery app and the reason is that it is very difficult to be sure that some scenario you believe it is edge case is really that rare to happen

Reply:
I agree that queuing a task while changing the celery_app should not happen, because it will probably make that task unrecoverable, but it's a tradeoff between speed in the 99% of the cases where the nodes don't have connectivity issues and faster recovery in the 1% of the cases where the node restarted.

Locks are not cheap and I don't think we should sacrifice the speed in EVERY call to the nodes, in order to gain a tiny bit faster recovery when a node failed (the scenario is that a queue task happens after an error signal is received and before the new app is set).

Reply:
I m not sure what you mean "locks are not cheap". Instantiating a lock is cheap. Acquiring and releasing a lock is cheap. What is not cheap is having multiple threads waiting to acquire the same lock, but this bottleneck is intentional, otherwise you would not lock the block. Calling `lock.acquire()` on a **released** lock is trivial timewise. When there are racing conditions (that are moreover not fully understood) you should use locks, not doing it for better speed performance, especially on a codebase where there are already multiple nondeterministic behaviors and fails, constitutes premature optimization and is plain wrong. 
When instantiating a new `Celery` object, any other part of the code that would make use of the `Celery` object should be blocked until the "exchange" of the `Celery` objects is complete, otherwise furtherdown the road it is very likely that you ll face this nondeterministic behavior and you ll have to revisit it again when instead you should make it robust now.

Reply:
This is not a scenario that a simple lock will suffice because we will have all threads competing for that lock in order to queue a task (that includes remote connection with rabbitmq and will be expensive and slow).

This is a scenario with multiple readers and one exclusive writer. Multiple threads can read the `celery_app` but only one can change the `celery_app` and while doing that exclude all readers. A python lock like that doesn't exist in python but can be implemented like this: https://www.oreilly.com/library/view/python-cookbook/0596001673/ch06s04.html

This though shows why it's going to be slow as well since all the time many readers will have to synchronously increase and decrease the readers count. That's what I meant by its not cheap.


I don't know what you refer to when you say "When there are racing conditions (that are moreover not fully understood)". The race conditions are fully understood and accepted, as a tradeoff for performance, due to their minor importance and rarity to occur. Also to clarify here that by race condition I don't mean that the system will crash or become unresponsive, just that untill the app is reset after we received the error, the tasks will keep failing in the meantime (much less than 1 second). If you want me to clarify furhtermore what these race conditions mean, let me know.


In the "race condition" where we close the app and then its used by a queue task, if the celery app changes it's behavior and actually closes the app making it unusable, then the queue task would simply throw an error or something similar I guess... This is the same scenario as in the `get_result`, since in the `get_result` the celery app is inside the result, so we can't do anything about that anyway, unless we want to lock the `get_result` as well and change the app somehow?

I could add a test to assert that closing an app doesn't do anything or remove that line from the codebase or both. That way if something changes in celery we will know by the failing test and adjust the code appropriately.

Reply:
Anyhow, I think the block that instantiates a new `Celery` object  in `_reset_celery_app` should be implemented in a way that allows only the first thread that gets in there to execute the block and ignore any subsequent thread from running the same block(instantiating a new `Celery` object), otherwise although the `Celery` object needs to be reinstantiated only once, it is completely random how many times it will be reset before everything "settles" down again.

Reply:
How would we know when the `get_result` call will happen? The `get_result` call could happen much later.